### ISSUE TITLE
Off-canvas legends

### ISSUE THREAD

**Issue by birdsarah** [bokeh/bokeh#2865]
We get asked a lot about moving legends off the plot.

At the moment there's no way to do that.

Looking at matplotlib & ggplot2:

* matplotlib: perhaps not as quick & easy as it should be: http://stackoverflow.com/questions/4700614/how-to-put-the-legend-out-of-the-plot#4701285
* ggplot2: the default left, right, top, bottom are off canvas but you can specify a position to have it on canvas too. http://docs.ggplot2.org/current/theme.html (about half way down the page)

As an additional requirement, I can also imagine, as a web person, wanting broader control over where a legend goes on your page.

For Bokeh, I think there's short-term and long-term solutions.  In the short-term maybe it would be nice to have the option to get a legend as plot object that you could at least pass to a `gridplot` or `components` method and then put it where you want. (maybe this is already possible and I haven't tried yet)

**Comment by VelizarVESSELINOV** [comment]
:+1: 
An old hope:
http://stackoverflow.com/questions/26254619/position-of-the-legend-in-a-bokeh-plot

**Comment by bryevdv** [comment]
There are to different use-cases, I just want to make sure to draw out the distinction:

*  A legend outside the main plot area, but still on the same canvas. 

* A legend on a completely different canvas, positionable as a DOM element.

Both of these are waiting on some work to clean up, update, and expand the low level canvas layout code, which will be a priority after `0.11`

**Comment by VelizarVESSELINOV** [comment]
For me ```A legend outside the main plot area, but still on the same canvas.``` is covering very well my current use-cases.

**Comment by bryevdv** [comment]
Happily, that's probably then one that will happen first. :)

**Comment by kynan** [comment]
Are there plans to support this for the charts API? Or is there a way of doing that already? I failed adapting the [example from the user guide](http://bokeh.pydata.org/en/latest/docs/user_guide/styling.html#outside-the-plot-area).

**Comment by birdsarah** [comment]
@kynan - I would suggest opening a new issue. It is possible, but it would be very hacky.

**Comment by kannanse** [comment]
@birdsarah - Can you show an example of doing it the hacky way? Has #4941 been implemented??

I tried chart.legend.location = (0.0,-30.0), but seems to remove the legend completely.

**Comment by antoniomolram** [comment]
as @kannanse says, it would be really helpful to have the feature as he said. It is much more clean as to create the Legend on a separate structure! ;)
